cpp/src/parquet/arrow/path_internal_test.cc

```diff
@@ -0,0 +1,426 @@
+  ASSERT_THAT(result.post_list_elements, SizeIs(1));
+  EXPECT_THAT(result.post_list_elements[0].start, Eq(0));
+  EXPECT_THAT(result.post_list_elements[0].end, Eq(4));
+}
+
+TEST_F(MultipathLevelBuilderTest, NullableSingleListWithAllPresentEntries) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_type = list(entries);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([[], [], [1], [], [2, 3]])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>{1, 1, 3, 1, 3, 3},
+                     /*rep_levels=*/std::vector<int16_t>{0, 0, 0, 0, 0, 1});
+
+  ASSERT_THAT(result.post_list_elements, SizeIs(1));
+  // JSON construction appears to only create child array lazily
+  EXPECT_THAT(result.post_list_elements[0].start, Eq(0));
+  EXPECT_THAT(result.post_list_elements[0].end, Eq(3));
+}
+
+TEST_F(MultipathLevelBuilderTest, NullableSingleListWithAllEmptyEntries) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_type = list(entries);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([[], [], [], [], []])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>(/*count=*/5, 1),
+                     /*rep_levels=*/std::vector<int16_t>(/*count=*/5, 0));
+}
+
+TEST_F(MultipathLevelBuilderTest, NullableSingleListWithSomeNullEntriesAndSomeNullLists) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_type = list(entries);
+  auto array = ::arrow::ArrayFromJSON(
+      list_type, R"([null, [1 , 2, 3], [], [], null,  null, [4, 5], [null]])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+
+  result.CheckLevels(
+      /*def_levels=*/std::vector<int16_t>{0, 3, 3, 3, 1, 1, 0, 0, 3, 3, 2},
+      /*rep_levels=*/std::vector<int16_t>{0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0});
+}
+
+TEST_F(MultipathLevelBuilderTest, NestedListsWithSomeEntries) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_field = field("list", list(entries), /*nullable=*/true);
+  auto nested_list_type = list(list_field);
+
+  auto array = ::arrow::ArrayFromJSON(
+      nested_list_type, R"([null, [[1 , 2, 3], [4, 5]], [[], [], []], []])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>{0, 5, 5, 5, 5, 5, 3, 3, 3, 1},
```

Comment:
in case you want to leave as a comment

```
// d 1 optional group bag {
// d 2   repeated group item (List) {
// d 3     optional group bag {
// d 4       repeated group item (List) {
// d 5         optional int64 Entries;
//         }
//       }
//     }
//    }
```